src/mame/drivers/nes_vt09.cpp

```diff
@@ -0,0 +1,514 @@
+
+
+protected:
+	required_device<nes_vt02_vt03_soc_device> m_soc;
+};
+
+class nes_vt09_state : public nes_vt09_common_state
+{
+public:
+	nes_vt09_state(const machine_config& mconfig, device_type type, const char* tag) :
+		nes_vt09_common_state(mconfig, type, tag)
+	{ }
+
+	void nes_vt09(machine_config& config);
+	void nes_vt09_1mb(machine_config& config);
+	void nes_vt09_2mb(machine_config& config);
+	void nes_vt09_4mb(machine_config& config);
+	void nes_vt09_4mb_rasterhack(machine_config& config);
+	void nes_vt09_8mb(machine_config& config);
+	void nes_vt09_16mb(machine_config& config);
+
+private:
+};
+
+uint8_t nes_vt09_common_base_state::vt_rom_r(offs_t offset)
+{
+	return m_prgrom[offset];
+}
+
+void nes_vt09_common_base_state::vtspace_w(offs_t offset, uint8_t data)
+{
+	logerror("%s: vtspace_w %08x : %02x", machine().describe_context(), offset, data);
+}
+
+// VTxx can address 25-bit address space (32MB of ROM) so use maps with mirroring in depending on ROM size
+void nes_vt09_common_state::vt_external_space_map_32mbyte(address_map &map)
+{
+	map(0x0000000, 0x1ffffff).r(FUNC(nes_vt09_common_state::vt_rom_r));
+}
+
+void nes_vt09_common_state::vt_external_space_map_16mbyte(address_map &map)
+{
+	map(0x0000000, 0x0ffffff).mirror(0x1000000).r(FUNC(nes_vt09_common_state::vt_rom_r));
+}
+
+void nes_vt09_common_state::vt_external_space_map_8mbyte(address_map &map)
+{
+	map(0x0000000, 0x07fffff).mirror(0x1800000).r(FUNC(nes_vt09_common_state::vt_rom_r));
+}
+
+void nes_vt09_common_state::vt_external_space_map_4mbyte(address_map &map)
+{
+	map(0x0000000, 0x03fffff).mirror(0x1c00000).r(FUNC(nes_vt09_common_state::vt_rom_r));
+}
+
+void nes_vt09_common_state::vt_external_space_map_2mbyte(address_map &map)
+{
+	map(0x0000000, 0x01fffff).mirror(0x1e00000).r(FUNC(nes_vt09_common_state::vt_rom_r));
+}
+
+void nes_vt09_common_state::vt_external_space_map_1mbyte(address_map &map)
+{
+	map(0x0000000, 0x00fffff).mirror(0x1f00000).r(FUNC(nes_vt09_common_state::vt_rom_r));
+}
+
+void nes_vt09_common_state::vt_external_space_map_512kbyte(address_map &map)
+{
+	map(0x0000000, 0x007ffff).mirror(0x1f80000).r(FUNC(nes_vt09_common_state::vt_rom_r));
+}
+
+
+uint8_t nes_vt09_common_base_state::extrain_0_r()
+{
+	if (m_exin0)
+		return m_exin0->read();
+	else
+	{
+		logerror("%s: extrain_0_r (not hooked up)\n", machine().describe_context());
+	}
+	return 0x00;
+}
+
+uint8_t nes_vt09_common_base_state::extrain_1_r()
+{
+	if (m_exin1)
+		return m_exin1->read();
+	else
+	{
+		logerror("%s: extrain_1_r (not hooked up)\n", machine().describe_context());
+	}
+	return 0x00;
+}
+
+uint8_t nes_vt09_common_base_state::extrain_2_r()
+{
+	if (m_exin2)
+		return m_exin2->read();
+	else
+	{
+		logerror("%s: extrain_2_r (not hooked up)\n", machine().describe_context());
+	}
+	return 0x00;
+}
+
+uint8_t nes_vt09_common_base_state::extrain_3_r()
+{
+	if (m_exin3)
+		return m_exin3->read();
+	else
+	{
+		logerror("%s: extrain_3_r (not hooked up)\n", machine().describe_context());
```

Comment:
If you used an `optional_ioport_array` for the extra input ports, you could make a template for these and avoid needing four copies of the function.